std/range/package.d

```diff
@@ -8578,6 +8578,503 @@ if (isForwardRange!Source && hasLength!Source)
+                    _use = use;
+                    _len = len;
+                }
+
+                void decrease() { if (_len != 0) _len--; }
+
+                @property bool empty() { return _len == 0; }
+
+                private:
+                    size_t _len;
+                    bool _use;
+                alias _use this;
+            }
+
+        private:
+            Source _source;
+            Partitioners _partitioners;
+            size_t _offset;
+            UseMaxLength _maxLen;
+            bool _isEmpty;
+        }
+        else
+        {
+            import std.typecons : RefCounted;
+
+            private static struct Chunk
+            {
+                private RefCounted!Impl impl;
+
+                @property bool empty() { return impl.curSizeLeft == 0 || impl.source.empty; }
+                @property auto front() { return impl.source.front; }
+                void popFront()
+                {
+                    impl.curSizeLeft--;
+                    impl.source.popFront();
+                }
+            }
+
+            private static struct Impl
+            {
+                private Source source;
+                private Partitioners partitioners;
+                private size_t curSizeLeft;
+            }
+
+            private RefCounted!Impl impl;
+            // for when there are extra elements in the source range
+            private bool tookLastChunk;
+            private size_t offset;
+
+            private this(Source source, Partitioners partitioners)
+            {
+                tookLastChunk = false;
+                offset = 0;
+                impl = RefCounted!Impl(source, partitioners,
+                    source.empty ? 0 : (partitioners.empty ? 0 : partitioners.front));
+            }
+
+            @property bool empty() { return impl.source.empty || tookLastChunk; }
+            @property Chunk front()
+            {
+                return Chunk(impl);
+            }
+
+            void popFront()
+            {
+                if (impl.partitioners.empty)
+                {
+                    tookLastChunk = true;
+                    return;
+                }
+
+                impl.curSizeLeft -= impl.source.popFrontN(impl.curSizeLeft) - offset;
+                if (!impl.source.empty && !impl.partitioners.empty)
+                {
+                    static if (is(PartitionType == ByIndex))
+                        offset = impl.partitioners.front;
+                    impl.partitioners.popFront();
+                    if (impl.partitioners.empty)
+                        impl.curSizeLeft = -1;
+                    else
+                        impl.curSizeLeft = impl.partitioners.front - offset;
+                }
+            }
+        }
+    }
+
+    return Chunks(source, partitioners);
+}
+
+/// The Partitioners range contains the desired lengths of the chunks
+struct ByLength;
+/// The Partitioners range contains the end indexes of each chunk
+struct ByIndex;
```

Comment:
Maybe these should documented enum values? (currently the style tester sees them as undocumented/untested public symbols)